### ISSUE TITLE
about PicklingError: Can't pickle <class 'toil.BotoCredentialAdapter'>

### ISSUE THREAD

**Issue by apollonfq** [ComparativeGenomicsToolkit/cactus#100]
Hi All:
I am doing genome analysis using toil/cactus. I built the HPC environment  among AWS us-west-1 region .  I could use the sample genome data that cactus provides to do the analysis successfully.  But when I used my own genome data , the calculation was failed, and cactus said :” “PicklingError: Can't pickle <class 'toil.BotoCredentialAdapter'>: it's not found as toil.BotoCredentialAdapter,  ERROR:toil.worker:Exiting the worker because of a failed job on host ip-172-31-13-145.us-west-1.compute.internal”  ”.  Do you have any advice on this ?(The attached is the detailed error log from one of worker nodes
[BlastSequencesAgainstEachOther_d278459b--12fa--490b--850c--973b2cbe16cd002.log](https://github.com/ComparativeGenomicsToolkit/cactus/files/3708047/BlastSequencesAgainstEachOther_d278459b--12fa--490b--850c--973b2cbe16cd002.log)
)

**Comment by Secretloong** [comment]
Hi apollonfq

I had also met this bug. I think there is something wrong between Cactus and Toil. All the errors were made by the default toil image from quay.io/ucsc_cgl (3.20 in my runing). If you use the original developer's docker container, it would work well. 

Add following parameters, Cactus will work.
`TOIL_DOCKER_REGISTRY=quay.io/jorlarmstrong` `TOIL_APPLIANCE_SELF=quay.io/joelarmstrong/toil:latest`

Hope the developers would find the reason of this bug.

Best

**Comment by brantfaircloth** [comment]
I have also run up against the same bug.  Checking workaround.

**Comment by brantfaircloth** [comment]
Workaround seems to lead to additional failures w/ current version of cactus and this toil appliance.

**Comment by diekhans** [comment]
we are able to reproduce this problem and debugging it.

**Comment by brantfaircloth** [comment]
Sounds good.  Happy to help test fixes - I looked through the code and wondered if this was a missing `.rv()` somewhere but was not able to track down.

**Comment by adamnovak** [comment]
The error is:

PicklingError: Can't pickle <class 'toil.BotoCredentialAdapter'>: it's not found as toil.BotoCredentialAdapter

?

That would be referring to the class I added to patch boto so we can use the boto3 credentials with it. Unless Cactus is trying to do something weird like pickle an open S3 connection I don't see why it would want to pickle one of these.

But the reason it isn't pickle-able is that it is defined inside a function in Toil. The solution to this is to refactor Toil at https://github.com/DataBiosphere/toil/blob/8491a03a620f09f3e477405b5d4a5fabfe92c8bd/src/toil/__init__.py#L415 in order to define the class outside the function, and without using function locals in the class. I think that that shouldn't be too hard.

**Comment by diekhans** [comment]
thanks @adamnovak, it doesn't seem to make much sense to pickle an
S3 connect.  I bet there is a reference to an object that shouldn't
be there.

**Comment by joelarmstrong** [comment]
Maybe toil switched to using dill in more places? The C version of pickle is fairly broken in many ways -- see for example the times it has caused toil deadlicks because it doesn't release the GIL properly -- not sure if dill uses that internally. In a workaround for toil's slow serialization of child jobs [I ran into cPickle errors](https://github.com/ComparativeGenomicsToolkit/cactus/blob/88a8553b138164a781af2fa81c22763f3ae8590a/src/cactus/shared/common.py#L1177). Similarly these errors were for things not really being pickled. I needed to force toil to switch to Python pickle to get it to serialize. So that ugly-hack-on-top-of-ugly-hack is probably to blame; the error message may even be a red herring. We could get rid of ChildTreeJob, however that would add a delay of roughly 2-4 hours (if I remember properly) for each subproblem.

**Comment by brantfaircloth** [comment]
Hi everyone - thanks for your help so far. I wanted to follow-up and see if there was a potential workaround (or fix in testing) for this error?  Or, do you have any suggested alternatives (e.g. move to another platform from AWS)?  This is a blocking error for a couple of our projects.

I'm also happy to test fixes/workarounds as needed or suggested. I'm not sure I understand the interplay between AWS, boto, toil, and cactus sufficiently to try to implement a fix on my own.

Thanks again,
b

**Comment by brantfaircloth** [comment]
Following up again - is there a fix or is the workaround @glennhickey proposes https://github.com/DataBiosphere/toil/issues/2881 the best, current approach to dealing with the pickling error?

Thanks you,
b

**Comment by glennhickey** [comment]
That toil issue still seems open.  A simpler workaround that doesn't involve hacking source files is to revert to Toil version `3.20.0`

**Comment by brantfaircloth** [comment]
Right on 👍 - I've got `sys.maxsize` in place of `maxChildren` in a fork and will give that a shot to see if the fix also works for me.  If that causes problems, I'll revert to `3.20.20`.  Thx!

**Comment by brantfaircloth** [comment]
The new fix in toil 3.23.0 seems to have some sort of remaining issue (see https://github.com/DataBiosphere/toil/issues/2881), and I am also running into problems when reverting to toil 3.20.0 (seems like the initial pickling issue).  E.g., this error from a few minutes ago when running on AWS:

```
WARNING:toil.leader:e39039b6-7c3d-4281-b0d8-7265b973f672      File "/usr/local/lib/python2.7/dist-packages/dill/dill.py", line 1299, in save_type
WARNING:toil.leader:e39039b6-7c3d-4281-b0d8-7265b973f672        StockPickler.save_global(pickler, obj)
WARNING:toil.leader:e39039b6-7c3d-4281-b0d8-7265b973f672      File "/usr/lib/python2.7/pickle.py", line 754, in save_global
WARNING:toil.leader:e39039b6-7c3d-4281-b0d8-7265b973f672        (obj, module, name))
WARNING:toil.leader:e39039b6-7c3d-4281-b0d8-7265b973f672    PicklingError: Can't pickle <class 'toil.BotoCredentialAdapter'>: it's not found as toil.BotoCredentialAdapter
```

One thing I'm still not clear about - if I edit the source as @glennhickey suggests, do I also need to build an image from the edited source?  I've recently had issues w/ missing images in quay.io (e.g. at least as of this weekend, I need to pass `--latest` to `cactus` rather than running the current commit in `cactus` master because an image could not be found in quay for the most recent  commit in master).
